From 1d4c5db15ea1280b87c94e59d48c0c5685027bb9 Mon Sep 17 00:00:00 2001
From: Peter Yin <peter.yin@quantatw.com>
Date: Wed, 26 Nov 2025 15:54:08 +0800
Subject: [PATCH 500/500] ARM64: Aspeed2700: add a config for WDTA timer reload
 value

Add a config for WDTA timer reload, and set timer reload value if
WDTA is enabled.

Signed-off-by: Peter Yin <peter.yin@quantatw.com>
---
 arch/arm/mach-aspeed/ast2700/Kconfig    |  9 +++++++++
 arch/arm/mach-aspeed/ast2700/scu_info.c | 12 ++++++++++++
 2 files changed, 21 insertions(+)

diff --git a/arch/arm/mach-aspeed/ast2700/Kconfig b/arch/arm/mach-aspeed/ast2700/Kconfig
index 75c15814..10b7701d 100644
--- a/arch/arm/mach-aspeed/ast2700/Kconfig
+++ b/arch/arm/mach-aspeed/ast2700/Kconfig
@@ -33,6 +33,15 @@ config TARGET_AST2700_DCSCM
 
 endchoice
 
+config ASPEED_WDTA_TIMER_RELOAD
+	hex "Aspeed WDTA timer reload value"
+	depends on ASPEED_AST2700
+	default 0x0
+	help
+	  Set the reload value for the Aspeed WDTA timer.
+	  The unit is 1us (e.g., 1 second = 0xF4240).
+	  Set to 0x0 to leave the register unchanged (Default 40s).
+
 source "board/aspeed/evb_ast2700/Kconfig"
 source "board/aspeed/ast2700_dcscm/Kconfig"
 
diff --git a/arch/arm/mach-aspeed/ast2700/scu_info.c b/arch/arm/mach-aspeed/ast2700/scu_info.c
index be8aff39..616d9b26 100644
--- a/arch/arm/mach-aspeed/ast2700/scu_info.c
+++ b/arch/arm/mach-aspeed/ast2700/scu_info.c
@@ -10,6 +10,8 @@
 #include <asm/arch/scu_ast2700.h>
 #include <asm/arch/spi.h>
 
+static u32 aspeed_wdta_reload = CONFIG_ASPEED_WDTA_TIMER_RELOAD;
+
 int aspeed_get_mac_phy_interface(u8 num)
 {
 	return 0;
@@ -67,6 +69,16 @@ void aspeed_print_2nd_wdt_mode(void)
 			printf(", infinite ABR\n");
 		else
 			printf("\n");
+
+		if (aspeed_wdta_reload &&
+		    (readl(ASPEED_WDTA_BASE + 0x0c) & BIT(0))) {
+			writel(aspeed_wdta_reload, ASPEED_WDTA_BASE + 0x04);
+			writel(0x4755, ASPEED_WDTA_BASE + 0x08);
+			printf("Set WDTA timer reload value to 0x%08X\n",
+			       aspeed_wdta_reload);
+		} else {
+			printf("Keep WDTA timer reload value as init\n");
+		}
 	}
 }
 
-- 
2.43.0

